mesh/src/immersed_boundary/ImmersedBoundaryMesh.cpp

```diff
@@ -0,0 +1,2322 @@
+modification, are permitted provided that the following conditions are met:
+ * Redistributions of source code must retain the above copyright notice,
+   this list of conditions and the following disclaimer.
+ * Redistributions in binary form must reproduce the above copyright notice,
+   this list of conditions and the following disclaimer in the documentation
+   and/or other materials provided with the distribution.
+ * Neither the name of the University of Oxford nor the names of its
+   contributors may be used to endorse or promote products derived from this
+   software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
+GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
+OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+#include "ImmersedBoundaryMesh.hpp"
+
+#include <algorithm>
+#include <cmath>
+
+#include "ImmersedBoundaryEnumerations.hpp"
+#include "MeshUtilityFunctions.hpp"
+#include "RandomNumberGenerator.hpp"
+#include "UblasCustomFunctions.hpp"
+#include "Warnings.hpp"
+
+
+template <unsigned ELEMENT_DIM, unsigned SPACE_DIM>
+ImmersedBoundaryMesh<ELEMENT_DIM, SPACE_DIM>::ImmersedBoundaryMesh(std::vector<Node<SPACE_DIM>*> nodes,
+                                                                   std::vector<ImmersedBoundaryElement<ELEMENT_DIM, SPACE_DIM>*> elements,
+                                                                   std::vector<ImmersedBoundaryElement<ELEMENT_DIM - 1, SPACE_DIM>*> laminas,
+                                                                   unsigned numGridPtsX,
+                                                                   unsigned numGridPtsY)
+        : mNumGridPtsX(numGridPtsX),
+          mNumGridPtsY(numGridPtsY),
+          mCharacteristicNodeSpacing(DOUBLE_UNSET),
+          mElementDivisionSpacing(DOUBLE_UNSET),
+          mNeighbourDist(0.1),
+          mSummaryOfNodeLocations(0.0),
+          mCellRearrangementThreshold(0.05)
+{
+    // Clear mNodes and mElements
+    Clear();
+
+    switch (SPACE_DIM)
+    {
+        case 2:
+            m2dVelocityGrids.resize(extents[2][mNumGridPtsX][mNumGridPtsY]);
+            break;
+
+        case 3:
+            EXCEPTION("Not implemented yet in 3D");
+            break;
+        
+        //LCOV_EXCL_START
+        default:
```

Comment:
If the 1D immersed boundary mesh never exists, do we need all the 1D template instantiations for all the immersed boundary stuff??